File: src/services/tools/web/webSearch.ts
import { z } from 'zod';
import type { ToolDef } from '../types';
import { WebSearchService } from '../../WebSearchService';

export const webSearchTool: ToolDef = {
  name: 'web_search',
  description: 'Search the web for current information. Use when you need up-to-date data, recent news, documentation, or facts beyond your training cutoff.',
  category: 'web',
  inputSchema: z.object({
    query: z.string().describe('The search query. Be specific and concise. Supports site:example.com to limit to one domain.'),
    site: z.string().optional().describe('Limit search to a single domain only (e.g. "react.dev").'),
    maxResults: z.number().min(1).max(5).optional().default(5).describe('Number of results (1–5).'),
  }),
  execute: async ({ query, site, maxResults }) => {
    try {
      const result = await WebSearchService.search({ query, site, maxResults: Math.min(maxResults ?? 5, 5) });
      return {
        results: result.results.map(r => ({
          title: r.title, url: r.url, snippet: r.snippet,
          content: r.content, publishedDate: r.publishedDate, source: r.source,
        })),
        totalResults: result.totalResults,
        answer: result.answer,
      };
    } catch (error: any) {
      throw new Error(error.message || 'Search request failed.');
    }
  },
};

File: agent/internal/tool/web_search.go
package tool

import (
	"context"
	"fmt"

	"github.com/xz-ephyr/raw-code/agent/pkg/api"
)

func webSearchTool() ToolDef {
	return ToolDef{
		Definition: api.ToolDefinition{
			Name:        "web_search",
			Description: "Search the web for up-to-date information on any topic. Use this when you need current information, news, documentation, or facts.",
			Category:    "research",
			Parameters: map[string]api.ParamDef{
				"query":       {Type: "string", Description: "The search query", Required: true},
				"max_results": {Type: "number", Description: "Max results (1-10)", Required: false, Default: 5},
				"site":        {Type: "string", Description: "Restrict search to a specific domain (e.g. developer.mozilla.org)", Required: false},
			},
		},
		Handler: func(ctx context.Context, e *Executor, params map[string]any) (any, error) {
			query, _ := params["query"].(string)
			if query == "" {
				return nil, fmt.Errorf("query is required")
			}
			maxResults, _ := params["max_results"].(float64)
			if maxResults == 0 {
				maxResults = 5
			}

			reqParams := map[string]any{
				"query":      query,
				"maxResults": int(maxResults),
			}
			if site, ok := params["site"].(string); ok && site != "" {
				reqParams["site"] = site
			}

			return postToExpress(ctx, e.expressURL, "webSearch", reqParams)
		},
	}
}
